File: scripts/sync.js
#!/usr/bin/env node
// Generate Tashtit's Codex adapters from their canonical sources.
//
// Codex requires two files that cannot live at the shared Claude/Copilot
// paths:
//
// - `.agents/plugins/marketplace.json`, whose catalog schema differs;
// - `.codex-plugin/plugin.json` per plugin, whose content matches the shared
//   manifest but whose location is fixed by the host.
//
// Both are generated here and drift-checked by `make validate`. Repository
// symlinks are deliberately not used: Git materializes them as plain text
// files when `core.symlinks=false`, which is the default whenever a checkout
// cannot create symlinks, and a host would then read the link target instead
// of JSON.

import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath, pathToFileURL } from 'node:url';

let ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

const SHARED_MANIFEST_DIR = '.claude-plugin';
const CODEX_MANIFEST_DIR = '.codex-plugin';
const DISPLAY_NAME = 'Tashtit';
const PLUGIN_NAME = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const PLATFORMS = new Set(['claude-code', 'codex', 'cursor', 'github-copilot']);
// Platforms a plugin targets by default when it declares no `platforms` list.
// Cursor is optional research, so it is opt-in rather than a default target.
const CORE_TARGETS = new Set(['claude-code', 'codex', 'github-copilot']);
const SEMVER =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/;
const REQUIRED_MANIFEST_FIELDS = ['name', 'version', 'description', 'skills'];

// Raised when a canonical source cannot be trusted for generation.
export class SyncError extends Error {}

export function setRoot(newRoot) {
  const previous = ROOT;
  ROOT = newRoot;
  return previous;
}

const sharedMarketplacePath = () =>
  path.join(ROOT, '.claude-plugin', 'marketplace.json');
const codexMarketplacePath = () =>
  path.join(ROOT, '.agents', 'plugins', 'marketplace.json');
const pluginsRoot = () => path.join(ROOT, 'plugins');

function relative(filePath) {
  const rel = path.relative(ROOT, filePath);
  return rel === '' || rel.startsWith('..') || path.isAbsolute(rel)
    ? String(filePath)
    : rel;
}

const isObjectValue = (value) =>
  typeof value === 'object' && value !== null && !Array.isArray(value);

function requireObject(value, field) {
  if (!isObjectValue(value)) {
    throw new SyncError(`${field} must be an object`);
  }
  return value;
}

function requireText(value, field) {
  if (typeof value !== 'string' || !value.trim()) {
    throw new SyncError(`${field} must be a non-empty string`);
  }
  return value;
}

function readJson(filePath) {
  let raw;
  try {
    raw = fs.readFileSync(filePath, 'utf8');
  } catch (error) {
    if (error.code === 'ENOENT') {
      throw new SyncError(`missing canonical source: ${relative(filePath)}`);
    }
    throw error;
  }
  try {
    return JSON.parse(raw);
  } catch (error) {
    throw new SyncError(`${relative(filePath)}: invalid JSON: ${error.message}`);
  }
}

export function loadSharedMarketplace() {
  const marketplace = requireObject(
    readJson(sharedMarketplacePath()),
    'marketplace',
  );
  if (requireText(marketplace.name, 'name') !== 'tashtit') {
    throw new SyncError("name must be 'tashtit'");
  }
  const owner = requireObject(marketplace.owner, 'owner');
  requireText(owner.name, 'owner.name');
  const metadata = requireObject(marketplace.metadata, 'metadata');
  requireText(metadata.description, 'metadata.description');
  if (!SEMVER.test(requireText(metadata.version, 'metadata.version'))) {
    throw new SyncError('metadata.version must use Semantic Versioning');
  }

  const plugins = marketplace.plugins;
  if (!Array.isArray(plugins)) {
    throw new SyncError('plugins must be an array');
  }

  const names = [];
  plugins.forEach((rawPlugin, index) => {
    const plugin = requireObject(rawPlugin, `plugins[${index}]`);
    const prefix = `plugins[${index}]`;
    const name = requireText(plugin.name, `${prefix}.name`);
    if (!PLUGIN_NAME.test(name)) {
      throw new SyncError(`${prefix}.name must use lowercase kebab-case`);
    }
    names.push(name);
    if (plugin.source !== `./plugins/${name}`) {
      throw new SyncError(`${prefix}.source must be './plugins/${name}'`);
    }
    requireText(plugin.description, `${prefix}.description`);
    if (!SEMVER.test(requireText(plugin.version, `${prefix}.version`))) {
      throw new SyncError(`${prefix}.version must use Semantic Versioning`);
    }
    if (plugin.license !== 'Apache-2.0') {
      throw new SyncError(`${prefix}.license must be 'Apache-2.0'`);
    }
    requireText(plugin.category, `${prefix}.category`);
    const platforms = plugin.platforms;
    if (platforms !== undefined && platforms !== null) {
      if (
        !Array.isArray(platforms) ||
        platforms.length === 0 ||
        platforms.some((item) => typeof item !== 'string')
      ) {
        throw new SyncError(
          `${prefix}.platforms must be a non-empty array of strings`,
        );
      }
      const unknown = platforms
        .filter((item) => !PLATFORMS.has(item))
        .sort();
      if (unknown.length > 0) {
        throw new SyncError(
          `${prefix}.platforms has unknown platforms ${JSON.stringify(unknown)}`,
        );
      }
    }
  });

  if (names.join('\n') !== [...names].sort().join('\n')) {
    throw new SyncError('plugins must be sorted by name');
  }
  if (names.length !== new Set(names).size) {
    throw new SyncError('plugin names must be unique');
  }
  return marketplace;
}

export function targetsCodex(entry) {
  // Whether a plugin gets a generated Codex adapter.
  //
  // A plugin targets every core platform unless it declares a narrower
  // `platforms` list. Codex is the only platform with a separately generated
  // adapter, so omitting it from `platforms` suppresses that adapter while
  // the shared Claude/Copilot manifest is unaffected.
  const platforms = entry.platforms;
  if (platforms === undefined || platforms === null) {
    return CORE_TARGETS.has('codex');
  }
  return platforms.includes('codex');
}

export function buildCodexMarketplace(marketplace) {
  // Translate the shared entries to Codex's incompatible catalog schema.
  const plugins = [];
  for (const plugin of marketplace.plugins) {
    if (!targetsCodex(plugin)) {
      continue;
    }
    plugins.push({
      name: plugin.name,
      source: {
        source: 'local',
        path: `./plugins/${plugin.name}`,
      },
      policy: {
        installation: 'AVAILABLE',
        authentication: 'ON_INSTALL',
      },
      category: plugin.category,
    });
  }
  return {
    name: marketplace.name,
    interface: { displayName: DISPLAY_NAME },
    plugins,
  };
}

function pluginDirectories() {
  let entries;
  try {
    entries = fs.readdirSync(pluginsRoot(), { withFileTypes: true });
  } catch {
    throw new SyncError(`missing plugins directory: ${relative(pluginsRoot())}`);
  }
  return entries
    .filter((entry) => entry.isDirectory() && !entry.name.startsWith('.'))
    .map((entry) => path.join(pluginsRoot(), entry.name))
    .sort();
}

export function loadSharedManifest(manifestPath, pluginName) {
  // Return the canonical manifest text once Codex's requirements hold.
  //
  // Codex accepts the shared manifest schema, so the generated file is a byte
  // copy. Validate here rather than emitting an adapter a host cannot load.
  const manifest = requireObject(
    readJson(manifestPath),
    `${relative(manifestPath)} manifest`,
  );
  if (manifest.name !== pluginName) {
    throw new SyncError(
      `${relative(manifestPath)}: name must be ${JSON.stringify(pluginName)}`,
    );
  }
  for (const field of REQUIRED_MANIFEST_FIELDS) {
    requireText(manifest[field], `${relative(manifestPath)}.${field}`);
  }
  if (!SEMVER.test(manifest.version)) {
    throw new SyncError(
      `${relative(manifestPath)}: version must use Semantic Versioning`,
    );
  }
  if (!manifest.skills.startsWith('./')) {
    throw new SyncError(
      `${relative(manifestPath)}: skills path must start with './'`,
    );
  }
  return fs.readFileSync(manifestPath, 'utf8');
}

export function render(value) {
  return JSON.stringify(value, null, 2) + '\n';
}

export function collectArtifacts() {
  // Pair every generated path with the content its canonical source implies.
  //
  // Also returns the generated adapters that must no longer exist, so that a
  // plugin dropping codex from its platforms is remediated by `make sync`
  // instead of leaving an adapter that only `make validate` complains about.
  const marketplace = loadSharedMarketplace();
  const codexTargets = new Set(
    marketplace.plugins
      .filter((plugin) => targetsCodex(plugin))
      .map((plugin) => plugin.name),
  );
  const artifacts = [
    [codexMarketplacePath(), render(buildCodexMarketplace(marketplace))],
  ];
  const obsolete = [];
  for (const pluginDir of pluginDirectories()) {
    const adapter = path.join(pluginDir, CODEX_MANIFEST_DIR, 'plugin.json');
    if (!codexTargets.has(path.basename(pluginDir))) {
      if (isFile(adapter) || isSymlink(adapter)) {
        obsolete.push(adapter);
      }
      continue;
    }
    const sharedManifest = path.join(
      pluginDir,
      SHARED_MANIFEST_DIR,
      'plugin.json',
    );
    artifacts.push([
      adapter,
      loadSharedManifest(sharedManifest, path.basename(pluginDir)),
    ]);
  }
  return { artifacts, obsolete };
}

export function removeObsolete(adapterPath, check) {
  // Remove one adapter whose plugin no longer targets codex.
  if (check) {
    process.stderr.write(
      `${relative(adapterPath)}: plugin no longer targets codex, so its ` +
        'generated adapter must be removed\n',
    );
    return false;
  }
  fs.unlinkSync(adapterPath);
  try {
    fs.rmdirSync(path.dirname(adapterPath));
  } catch {
    // The directory holds files this script does not generate; validation
    // will report them if they are a problem.
  }
  process.stdout.write(`removed ${relative(adapterPath)}\n`);
  return true;
}

function splitKeepingNewlines(text) {
  return text.split(/(?<=\n)/);
}

// Minimal unified-style diff: trims the common prefix and suffix and prints
// the differing middle. Enough to show which generated lines drifted.
function writeDiff(actual, expected, fromFile, toFile) {
  const actualLines = splitKeepingNewlines(actual);
  const expectedLines = splitKeepingNewlines(expected);

  let start = 0;
  while (
    start < actualLines.length &&
    start < expectedLines.length &&
    actualLines[start] === expectedLines[start]
  ) {
    start += 1;
  }
  let actualEnd = actualLines.length;
  let expectedEnd = expectedLines.length;
  while (
    actualEnd > start &&
    expectedEnd > start &&
    actualLines[actualEnd - 1] === expectedLines[expectedEnd - 1]
  ) {
    actualEnd -= 1;
    expectedEnd -= 1;
  }

  const ensureNewline = (line) => (line.endsWith('\n') ? line : `${line}\n`);
  process.stderr.write(`--- ${fromFile}\n`);
  process.stderr.write(`+++ ${toFile}\n`);
  for (const line of actualLines.slice(start, actualEnd)) {
    process.stderr.write(ensureNewline(`-${line.replace(/\n$/, '')}`));
  }
  for (const line of expectedLines.slice(start, expectedEnd)) {
    process.stderr.write(ensureNewline(`+${line.replace(/\n$/, '')}`));
  }
}

function isSymlink(filePath) {
  try {
    return fs.lstatSync(filePath).isSymbolicLink();
  } catch {
    return false;
  }
}

function isFile(filePath) {
  try {
    return fs.statSync(filePath).isFile();
  } catch {
    return false;
  }
}

export function syncArtifact(artifactPath, expected, check) {
  // Write or verify one generated file, replacing any symlink in place.
  const linked = isSymlink(artifactPath);
  const actual = isFile(artifactPath)
    ? fs.readFileSync(artifactPath, 'utf8')
    : '';
  if (!linked && actual === expected) {
    return true;
  }

  if (check) {
    if (linked) {
      process.stderr.write(
        `${relative(artifactPath)}: must be a generated regular file, not a ` +
          'symlink; symlinks become plain text when core.symlinks=false\n',
      );
      return false;
    }
    writeDiff(
      actual,
      expected,
      relative(artifactPath),
      `${relative(artifactPath)} (generated)`,
    );
    return false;
  }

  if (linked) {
    fs.unlinkSync(artifactPath);
  }
  fs.mkdirSync(path.dirname(artifactPath), { recursive: true });
  fs.writeFileSync(artifactPath, expected);
  process.stdout.write(`updated ${relative(artifactPath)}\n`);
  return true;
}

export function main() {
  const check = process.argv.includes('--check');

  let collected;
  try {
    collected = collectArtifacts();
  } catch (error) {
    if (error instanceof SyncError) {
      process.stderr.write(
        `canonical source validation failed: ${error.message}\n`,
      );
      return 1;
    }
    throw error;
  }

  const { artifacts, obsolete } = collected;
  const stale = artifacts.filter(
    ([artifactPath, expected]) => !syncArtifact(artifactPath, expected, check),
  );
  stale.push(
    ...obsolete.filter((adapterPath) => !removeObsolete(adapterPath, check)),
  );
  if (stale.length > 0) {
    process.stderr.write(
      'Generated Codex adapters are stale; run `make sync` and include ' +
        'the result\n',
    );
    return 1;
  }
  if (check) {
    process.stdout.write(
      `Generated Codex adapters are synchronized (${artifacts.length} files).\n`,
    );
  }
  return 0;
}

if (
  process.argv[1] &&
  import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href
) {
  process.exit(main());
}

File: scripts/sync.py
#!/usr/bin/env python3
"""Generate Tashtit's Codex adapters from their canonical sources.

Codex requires two files that cannot live at the shared Claude/Copilot paths:

- `.agents/plugins/marketplace.json`, whose catalog schema differs;
- `.codex-plugin/plugin.json` per plugin, whose content matches the shared
  manifest but whose location is fixed by the host.

Both are generated here and drift-checked by `make validate`. Repository
symlinks are deliberately not used: Git materializes them as plain text files
when `core.symlinks=false`, which is the default whenever a checkout cannot
create symlinks, and a host would then read the link target instead of JSON.
"""

from __future__ import annotations

import argparse
import difflib
import json
import re
import sys
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parent.parent
SHARED_MARKETPLACE_PATH = ROOT / ".claude-plugin" / "marketplace.json"
CODEX_MARKETPLACE_PATH = ROOT / ".agents" / "plugins" / "marketplace.json"
PLUGINS_ROOT = ROOT / "plugins"
SHARED_MANIFEST_DIR = ".claude-plugin"
CODEX_MANIFEST_DIR = ".codex-plugin"
DISPLAY_NAME = "Tashtit"
PLUGIN_NAME = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
PLATFORMS = {"claude-code", "codex", "cursor", "github-copilot"}
# Platforms a plugin targets by default when it declares no `platforms` list.
# Cursor is optional research, so it is opt-in rather than a default target.
CORE_TARGETS = {"claude-code", "codex", "github-copilot"}
SEMVER = re.compile(
    r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)"
    r"(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?"
    r"(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$"
)
REQUIRED_MANIFEST_FIELDS = ("name", "version", "description", "skills")


class SyncError(ValueError):
    """Raised when a canonical source cannot be trusted for generation."""


def relative(path: Path) -> str:
    try:
        return str(path.relative_to(ROOT))
    except ValueError:
        return str(path)


def require_object(value: Any, field: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise SyncError(f"{field} must be an object")
    return value


def require_text(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise SyncError(f"{field} must be a non-empty string")
    return value


def read_json(path: Path) -> Any:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError as error:
        raise SyncError(f"missing canonical source: {relative(path)}") from error
    except json.JSONDecodeError as error:
        raise SyncError(
            f"{relative(path)}: invalid JSON at line {error.lineno}, "
            f"column {error.colno}"
        ) from error


def load_shared_marketplace() -> dict[str, Any]:
    marketplace = require_object(read_json(SHARED_MARKETPLACE_PATH), "marketplace")
    if require_text(marketplace.get("name"), "name") != "tashtit":
        raise SyncError("name must be 'tashtit'")
    owner = require_object(marketplace.get("owner"), "owner")
    require_text(owner.get("name"), "owner.name")
    metadata = require_object(marketplace.get("metadata"), "metadata")
    require_text(metadata.get("description"), "metadata.description")
    if not SEMVER.fullmatch(
        require_text(metadata.get("version"), "metadata.version")
    ):
        raise SyncError("metadata.version must use Semantic Versioning")

    plugins = marketplace.get("plugins")
    if not isinstance(plugins, list):
        raise SyncError("plugins must be an array")

    names: list[str] = []
    for index, raw_plugin in enumerate(plugins):
        plugin = require_object(raw_plugin, f"plugins[{index}]")
        prefix = f"plugins[{index}]"
        name = require_text(plugin.get("name"), f"{prefix}.name")
        if not PLUGIN_NAME.fullmatch(name):
            raise SyncError(f"{prefix}.name must use lowercase kebab-case")
        names.append(name)
        if plugin.get("source") != f"./plugins/{name}":
            raise SyncError(f"{prefix}.source must be './plugins/{name}'")
        require_text(plugin.get("description"), f"{prefix}.description")
        if not SEMVER.fullmatch(
            require_text(plugin.get("version"), f"{prefix}.version")
        ):
            raise SyncError(f"{prefix}.version must use Semantic Versioning")
        if plugin.get("license") != "Apache-2.0":
            raise SyncError(f"{prefix}.license must be 'Apache-2.0'")
        require_text(plugin.get("category"), f"{prefix}.category")
        platforms = plugin.get("platforms")
        if platforms is not None:
            if (
                not isinstance(platforms, list)
                or not platforms
                or any(not isinstance(item, str) for item in platforms)
            ):
                raise SyncError(
                    f"{prefix}.platforms must be a non-empty array of strings"
                )
            unknown = sorted(set(platforms) - PLATFORMS)
            if unknown:
                raise SyncError(f"{prefix}.platforms has unknown platforms {unknown}")

    if names != sorted(names):
        raise SyncError("plugins must be sorted by name")
    if len(names) != len(set(names)):
        raise SyncError("plugin names must be unique")
    return marketplace


def targets_codex(entry: dict[str, Any]) -> bool:
    """Whether a plugin gets a generated Codex adapter.

    A plugin targets every core platform unless it declares a narrower
    `platforms` list. Codex is the only platform with a separately generated
    adapter, so omitting it from `platforms` suppresses that adapter while the
    shared Claude/Copilot manifest is unaffected.
    """
    platforms = entry.get("platforms")
    if platforms is None:
        return "codex" in CORE_TARGETS
    return "codex" in platforms


def build_codex_marketplace(marketplace: dict[str, Any]) -> dict[str, Any]:
    """Translate the shared entries to Codex's incompatible catalog schema."""
    plugins = []
    for plugin in marketplace["plugins"]:
        if not targets_codex(plugin):
            continue
        plugins.append(
            {
                "name": plugin["name"],
                "source": {
                    "source": "local",
                    "path": f"./plugins/{plugin['name']}",
                },
                "policy": {
                    "installation": "AVAILABLE",
                    "authentication": "ON_INSTALL",
                },
                "category": plugin["category"],
            }
        )
    return {
        "name": marketplace["name"],
        "interface": {"displayName": DISPLAY_NAME},
        "plugins": plugins,
    }


def plugin_directories() -> list[Path]:
    if not PLUGINS_ROOT.is_dir():
        raise SyncError(f"missing plugins directory: {relative(PLUGINS_ROOT)}")
    return sorted(
        path
        for path in PLUGINS_ROOT.iterdir()
        if path.is_dir() and not path.name.startswith(".")
    )


def load_shared_manifest(path: Path, plugin_name: str) -> str:
    """Return the canonical manifest text once Codex's requirements hold.

    Codex accepts the shared manifest schema, so the generated file is a byte
    copy. Validate here rather than emitting an adapter a host cannot load.
    """
    manifest = require_object(read_json(path), f"{relative(path)} manifest")
    if manifest.get("name") != plugin_name:
        raise SyncError(f"{relative(path)}: name must be {plugin_name!r}")
    for field in REQUIRED_MANIFEST_FIELDS:
        require_text(manifest.get(field), f"{relative(path)}.{field}")
    if not SEMVER.fullmatch(manifest["version"]):
        raise SyncError(f"{relative(path)}: version must use Semantic Versioning")
    if not manifest["skills"].startswith("./"):
        raise SyncError(f"{relative(path)}: skills path must start with './'")
    return path.read_text(encoding="utf-8")


def render(value: dict[str, Any]) -> str:
    return json.dumps(value, indent=2, ensure_ascii=False) + "\n"


def collect_artifacts() -> tuple[list[tuple[Path, str]], list[Path]]:
    """Pair every generated path with the content its canonical source implies.

    Also returns the generated adapters that must no longer exist, so that a
    plugin dropping codex from its platforms is remediated by `make sync`
    instead of leaving an adapter that only `make validate` complains about.
    """
    marketplace = load_shared_marketplace()
    codex_targets = {
        plugin["name"]
        for plugin in marketplace["plugins"]
        if targets_codex(plugin)
    }
    artifacts = [
        (CODEX_MARKETPLACE_PATH, render(build_codex_marketplace(marketplace)))
    ]
    obsolete: list[Path] = []
    for plugin_dir in plugin_directories():
        adapter = plugin_dir / CODEX_MANIFEST_DIR / "plugin.json"
        if plugin_dir.name not in codex_targets:
            if adapter.is_file() or adapter.is_symlink():
                obsolete.append(adapter)
            continue
        shared_manifest = plugin_dir / SHARED_MANIFEST_DIR / "plugin.json"
        artifacts.append(
            (adapter, load_shared_manifest(shared_manifest, plugin_dir.name))
        )
    return artifacts, obsolete


def sync_artifact(path: Path, expected: str, check: bool) -> bool:
    """Write or verify one generated file, replacing any symlink in place."""
    linked = path.is_symlink()
    actual = path.read_text(encoding="utf-8") if path.is_file() else ""
    if not linked and actual == expected:
        return True

    if check:
        if linked:
            print(
                f"{relative(path)}: must be a generated regular file, not a "
                "symlink; symlinks become plain text when core.symlinks=false",
                file=sys.stderr,
            )
            return False
        sys.stderr.writelines(
            difflib.unified_diff(
                actual.splitlines(keepends=True),
                expected.splitlines(keepends=True),
                fromfile=relative(path),
                tofile=f"{relative(path)} (generated)",
            )
        )
        return False

    if linked:
        path.unlink()
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(expected, encoding="utf-8")
    print(f"updated {relative(path)}")
    return True


def remove_obsolete(path: Path, check: bool) -> bool:
    """Remove one adapter whose plugin no longer targets codex."""
    if check:
        print(
            f"{relative(path)}: plugin no longer targets codex, so its "
            "generated adapter must be removed",
            file=sys.stderr,
        )
        return False
    path.unlink()
    try:
        path.parent.rmdir()
    except OSError:
        # The directory holds files this script does not generate; validation
        # will report them if they are a problem.
        pass
    print(f"removed {relative(path)}")
    return True


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--check",
        action="store_true",
        help="fail if a generated Codex adapter is out of date",
    )
    args = parser.parse_args()

    try:
        artifacts, obsolete = collect_artifacts()
    except SyncError as error:
        print(f"canonical source validation failed: {error}", file=sys.stderr)
        return 1

    stale = [
        path
        for path, expected in artifacts
        if not sync_artifact(path, expected, args.check)
    ]
    stale.extend(
        path for path in obsolete if not remove_obsolete(path, args.check)
    )
    if stale:
        print(
            "Generated Codex adapters are stale; run `make sync` and include "
            "the result",
            file=sys.stderr,
        )
        return 1
    if args.check:
        print(f"Generated Codex adapters are synchronized ({len(artifacts)} files).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
